clientupdate/distsign/distsign.go

```diff
@@ -0,0 +1,231 @@
+// download and can be rotated more readily, assuming that most deployed
+// clients trust the root keys used to new fresh signing keys.
+package distsign
+
+import (
+	"crypto/ed25519"
+	"crypto/rand"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+)
+
+const (
+	pemTypePrivate = "PRIVATE KEY"
+	pemTypePublic  = "PUBLIC KEY"
+)
+
+// GenerateKey generates a new key pair and encodes it as PEM.
+func GenerateKey() (priv, pub []byte, retErr error) {
+	pub, priv, err := ed25519.GenerateKey(rand.Reader)
+	if err != nil {
+		return nil, nil, err
+	}
+	return pem.EncodeToMemory(&pem.Block{
+			Type:  pemTypePrivate,
+			Bytes: []byte(priv),
+		}), pem.EncodeToMemory(&pem.Block{
+			Type:  pemTypePublic,
+			Bytes: []byte(pub),
+		}), nil
+}
+
+// Signer is a single key (root or signing) that can sign arbitrary blobs.
+type Signer struct {
+	pk ed25519.PrivateKey
+}
+
+// NewSigner parses the PEM-encoded private key and creates a Signer for it.
+// The key is expected to be in the same format as returned by GenerateKey.
+func NewSigner(privKeyPath string) (*Signer, error) {
+	raw, err := os.ReadFile(privKeyPath)
+	if err != nil {
+		return nil, err
+	}
+	k, err := parsePrivateKey(raw)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse %q", privKeyPath)
+	}
+	return &Signer{pk: k}, nil
+}
+
+// Sign signs msg and returns the signature.
+func (s *Signer) Sign(msg []byte) []byte {
+	return ed25519.Sign(s.pk, msg)
+}
+
+// Client downloads and validates files from a distribution server.
+type Client struct {
+	roots    []ed25519.PublicKey
+	pkgsAddr string
+}
+
+// NewClient returns a new client for distribution server located at pkgsAddr,
+// and uses embedded root keys from the roots/ subdirectory of this package.
+func NewClient(pkgsAddr string) (*Client, error) {
+	return &Client{roots: roots(), pkgsAddr: pkgsAddr}, nil
+}
+
+// Download fetches a file at path srcPath from pkgsAddr passed in NewClient.
+// The file is downloaded to dstPath and its signature is validated using the
+// embedded root keys. Download returns an error if anything goes wrong with
+// the actual file download or with signature validation.
+func (c *Client) Download(srcPath, dstPath string) error {
+	// TODO(awly): don't buffer in memory
+
+	// Always fetch a fresh signing key.
+	sigPub, err := c.signingKeys()
+	if err != nil {
+		return err
+	}
+
+	srcURL := fmt.Sprintf("%s/%s", c.pkgsAddr, srcPath)
+	sigURL := srcURL + ".sig"
+
+	data, err := download(srcURL, 1<<30) // limit to 1Gb.
+	if err != nil {
+		return err
+	}
+	sig, err := download(sigURL, ed25519.SignatureSize)
+	if err != nil {
+		return err
+	}
+	if !verifyAny(sigPub, data, sig) {
+		return fmt.Errorf("signature %q for key %q does not validate with the current release signing key; either you are under attack, or attempting to download an old version of Tailscale which was signed with an older signing key", srcURL, sigURL)
```

Comment:
If I'm reading this right (might not be), this should be:

```
  return fmt.Errorf("signature %q for package %q does not...", sigURL, srcURL)
```

Reply:
oops, fixed!